cloud/scope/managedcontrolplane.go

```diff
@@ -0,0 +1,205 @@
+		patchHelper:            helper,
+	}, nil
+}
+
+// ManagedControlPlaneScope defines the basic context for an actuator to operate upon.
+type ManagedControlPlaneScope struct {
+	client      client.Client
+	patchHelper *patch.Helper
+
+	Cluster                *clusterv1.Cluster
+	GCPManagedCluster      *infrav1exp.GCPManagedCluster
+	GCPManagedControlPlane *infrav1exp.GCPManagedControlPlane
+	mcClient               *container.ClusterManagerClient
+	credentialsClient      *credentials.IamCredentialsClient
+	credential             *Credential
+}
+
+// PatchObject persists the managed control plane configuration and status.
+func (s *ManagedControlPlaneScope) PatchObject() error {
+	return s.patchHelper.Patch(
+		context.TODO(),
+		s.GCPManagedControlPlane,
+		patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{
+			infrav1exp.GKEControlPlaneReadyCondition,
+			infrav1exp.GKEControlPlaneCreatingCondition,
+			infrav1exp.GKEControlPlaneUpdatingCondition,
+			infrav1exp.GKEControlPlaneDeletingCondition,
+		}})
+}
+
+// Close closes the current scope persisting the managed control plane configuration and status.
+func (s *ManagedControlPlaneScope) Close() error {
+	s.mcClient.Close()
+	s.credentialsClient.Close()
+	return s.PatchObject()
+}
+
+// ConditionSetter return a condition setter (which is GCPManagedControlPlane itself).
+func (s *ManagedControlPlaneScope) ConditionSetter() conditions.Setter {
+	return s.GCPManagedControlPlane
+}
+
+// Client returns a k8s client.
+func (s *ManagedControlPlaneScope) Client() client.Client {
+	return s.client
+}
+
+// ManagedControlPlaneClient returns a client used to interact with GKE.
+func (s *ManagedControlPlaneScope) ManagedControlPlaneClient() *container.ClusterManagerClient {
+	return s.mcClient
+}
+
+// CredentialsClient returns a client used to interact with IAM.
+func (s *ManagedControlPlaneScope) CredentialsClient() *credentials.IamCredentialsClient {
+	return s.credentialsClient
+}
+
+// GetCredential returns the credential data.
+func (s *ManagedControlPlaneScope) GetCredential() *Credential {
+	return s.credential
+}
+
+func parseLocation(location string) (region string, zone *string) {
+	parts := strings.Split(location, "-")
+	region = strings.Join(parts[:2], "-")
+	if len(parts) == 3 {
+		return region, &parts[2]
+	}
+	return region, nil
+}
+
+// Region returns the region of the GKE cluster.
+func (s *ManagedControlPlaneScope) Region() string {
+	region, _ := parseLocation(s.GCPManagedControlPlane.Spec.Location)
+	return region
+}
+
+// ClusterLocation returns the location of the cluster.
+func (s *ManagedControlPlaneScope) ClusterLocation() string {
+	return fmt.Sprintf("projects/%s/locations/%s", s.GCPManagedControlPlane.Spec.Project, s.Region())
+}
+
+// ClusterFullName returns the full name of the cluster.
+func (s *ManagedControlPlaneScope) ClusterFullName() string {
+	return fmt.Sprintf("%s/clusters/%s", s.ClusterLocation(), s.GCPManagedControlPlane.Name)
```

Comment:
Should we also include the k8s namespace in the cluster name? We do this in EKS to reduce the potential for name collisions with cluster definitions in different namespaces named the same and using the same GCP project.

Reply:
great suggestion, updated the logic so user can specify cluster name, and CPAG will default to a cluster name with k8s namespace in it if that's not specified